crates/flox-rust-sdk/src/models/environment.rs

```diff
@@ -1,83 +1,764 @@
+        self.state.environment_ref()
     }
 
-    pub fn system(&self) -> Cow<str> {
-        match self {
-            CommonEnvironment::Named(n) => n.system(),
-            CommonEnvironment::Project(p) => p.system(),
+    /// Turn the environment into a flake attribute,
+    /// a precise url to interact with the environment via nix
+    ///
+    /// ```
+    /// # tokio_test::block_on(async {
+    /// # use flox_rust_sdk::models::environment::{Environment, Read};
+    /// # use std::path::PathBuf;
+    /// # let tempdir = tempfile::tempdir().unwrap();
+    /// # let path = tempdir.path().canonicalize().unwrap().to_string_lossy().into_owned();
+    /// # let system = "aarch64-darwin";
+    ///
+    /// let env = Environment::init(&path, "test_env".parse().unwrap())
+    ///     .await
+    ///     .unwrap();
+    ///
+    /// let flake_attribute = format!("path:{path}/.flox/test_env#.floxEnvs.{system}.default")
+    ///     .parse()
+    ///     .unwrap();
+    /// assert_eq!(env.flake_attribute(system), flake_attribute)
+    ///
+    /// # })
+    /// ```
+    pub fn flake_attribute(&self, system: impl AsRef<str>) -> FlakeAttribute {
+        let flakeref = PathRef {
+            path: self.path.clone(),
+            attributes: Default::default(),
+        }
+        .into();
+
+        let attr_path = ["", "floxEnvs", system.as_ref(), "default"]
+            .try_into()
+            .unwrap(); // validated attributes
+
+        FlakeAttribute {
+            flakeref,
+            attr_path,
+            outputs: Default::default(),
         }
     }
 
-    pub fn packages(&self) -> Value {
-        todo!()
+    /// Path to the environment definition file
+    pub fn flox_nix_path(&self) -> PathBuf {
+        self.path.join("pkgs").join("default").join("flox.nix")
     }
 
-    pub async fn install(&self, _packages: &[FloxPackage]) {
-        todo!()
+    /// Build the evironment derivation using nix
+    async fn build(
+        &self,
+        nix: &NixCommandLine,
+        system: impl AsRef<str>,
+    ) -> Result<(), EnvironmentError2> {
+        debug!("building with nix ....");
+
+        let build = Build {
+            installables: [self.flake_attribute(system).into()].into(),
+            eval: runix::arguments::eval::EvaluationArgs {
+                impure: true.into(),
+                ..Default::default()
+            },
+            build: BuildArgs {
+                // out_link: Some(self.out_link().into()),
+                ..Default::default()
+            },
+            ..Default::default()
+        };
+
+        build
+            .run(nix, &Default::default())
+            .await
+            .map_err(EnvironmentError2::Build)?;
+
+        Ok(())
     }
 
-    pub async fn uninstall(&self, _packages: &[FloxPackage]) {
-        todo!()
+    /// Get a catalog of installed packages from the environment
+    ///
+    /// Evaluated using nix from the environment definition.
+    pub async fn catalog(
```

Comment:
I'm confused, I thought the catalog kept track of which packages were built or provided by an environment. This says "installed from the environment". Is this method returning a list of packages _built from this environment_ that are also _installed in this environment_?

Reply:
This seems to be performing an eval to determine what catalog entries there are in the flake.

Reply:
Then I would advocate for a very explicit name like `catalog_entries_provided` or something similar. We aren't paying per character and we should strive to make it impossible to misunderstand what the method is doing.